pointcloud/octree/octree.go

```diff
@@ -0,0 +1,58 @@
+// Package octree implements a octree representation of pointclouds for easy traversal and storage of
+// probability and color data
+package octree
+
+import (
+	"github.com/golang/geo/r3"
+
+	pc "go.viam.com/rdk/pointcloud"
+)
+
+// Each node in the octree is either an internal node which links to other nodes, is an empty node with
+// no points or further links, or is an occupied node which contains a single point of data that
+// includes location, color and probability information.
+const (
+	InternalNode = NodeType(iota)
+	LeafNodeEmpty
+	LeafNodeFilled
+	octreeVersion = 1.0
+)
+
+// NodeType represents the types of nodes possible in an octree.
+type NodeType uint8
+
+// Octree is a data structure recursively partitions 3D space into octants to represent occupancy. It is
+// a storage format for a pointcloud that allows for better searchability and serialization. Each node is
+// either an internal node, empty node or child node. This implementation of an octree is compatible with
+// the pointcloud representation and includes a marshaling function.
+type Octree interface {
+	pc.PointCloud
+	Marshaler
+}
+
+// Marshaler will convert an octree into a serialized array of bytes.
+type Marshaler interface {
+	MarshalOctree() ([]byte, error)
+}
+
+// Unmarshaler will convert a serialized octree into an Octree datatype.
+type Unmarshaler interface {
+	UnmarshalOctree() (Octree, error)
+}
+
+// MetaData regarding data stored in the octree.
+type MetaData struct {
+	Version int32
+
+	CenterX, CenterY, CenterZ float64
+	Side                      float64
+	Size                      int32
```

Comment:
[q] Why is Size repeated here, when its already a method of the octree? If you need Size when serializing, you can just call your tree.Size(). It can be confusing having what is the same information being repeated in multiple places, and also creates the potential for bugs -- what if it is updated one place, but not the other?

Reply:
Can't you put size as a member field in basicOctree?
```
type basicOctree struct {
	logger golog.Logger
	node   basicOctreeNode
        size int 
	center r3.Vector
	side   float64
	meta   MetaData
}
```
And then calls to size can access the member field 
`func (octree *basicOctree) Size() int { return octree.size }` 

As well as when the metadata needs the size. It can just access it from the member field.

Reply:
If you don't want to make size a member field of octree, then having Size() return octree.meta.Size make most sense to me, as it keeps the function O(1)

Reply:
Solved, by removing octree metadata and placing size in basicOctree struct